app/src/main/java/org/mozilla/fenix/components/metrics/AppStartupTelemetry.kt

```diff
@@ -143,4 +157,17 @@ class AppStartupTelemetry(private val metrics: MetricController) : LifecycleObse
         // Therefore we set the isMetricRecorded flag to false.
         isMetricRecordedSinceAppWasForegrounded = false
     }
+
+    /**
+     * We record metric on onPreDraw() since we need to wait for the application to draw the
+     * first frame. which will always happen after onResume. The hot startup time is a bit complicated
+     * because if we have onPreDraw listener attached, Activity will invoke onPreDraw() before onRestart.
+     * Therefore, we use OneShotPreDrawListener, which removes the listener after first call.
+     */
+    fun onPreDraw() {
+        appLaunchTimeMeasurement.onActivityOnPreDraw()
+        MainScope().launch(Dispatchers.IO) {
```

Comment:
There's still a race – what if `onStop` clears this data before we're done in `recordMetric`?

I think the simplest solution is to transmit in `onStop` and block, assuming we don't expect it to take too long. A few other solutions:
- We can record here but we copy the data into the `recordMetric` function so that it can be cleared later without affecting the state. Note that we'll likely end up moving this later to minimize what runs before visual completeness
- We lock the data from here until we're done recording such that `onStop` can't clear it until we're done clearing (this is hard because we have to lock it before we even dispatch and even when we suspend to guarantee correctness... there might be a coroutines lock to help here)

Reply:
> transmit in onStop and block

guess simplest is to use `runBlocking`? or a coroutine on main thread. though I guess they both gets us same result at the end for this case?

Reply:
in the case where, `onstop` is called before `onPreDraw`, Im thinking of omiting the startup time but still report the other metrics?

Reply:
> are you thinking of using runBlockingor a coroutine on main thread? though I guess they both gets us same result at the end for this case?

I think we should get rid of the `suspend` keywords and just run it on the main thread. If you do `runBlocking` to a different dispatcher, you're potentially utilizing two threads (one will be sleeping though) and spending some resources dispatching. We might as well just run it on the main thread.

Since we're reading from the pseudo-filesystem and not the file system, I'm not too concerned about the performance of this (though we could measure locally or add a probe to measure the length just in case 🤷).

> in the case where, onstop is called before onPreDraw, Im thinking of omiting the startup time but still report the other metrics?

I'm not sure – in what cases can this happen? Does it change how our startup mental model should work?